drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1705 @@
+				    foo_data.data_len) {
+					esp8266_read_data(&rx_buf);
+
+					if (!rx_buf) {
+						break;
+					}
+					continue;
+				} else {
+					break;
+				}
+			}
+
+			net_buf_skipcrlf(&rx_buf);
+			if (!rx_buf) {
+				break;
+			}
+
+			/* check for incoming IPD data */
+			if (match(rx_buf, "+IPD") >= 0) {
+				if (esp8266_process_setup_read(&rx_buf) < 0) {
+					break;
+				}
+
+				if (!rx_buf) {
+					break;
+				}
+
+				continue;
+			}
+
+			if (match(rx_buf, ">") >= 0) {
+				rx_buf = net_buf_skip(rx_buf, 1);
+				k_sem_give(&foo_data.response_sem);
+			}
+
+			frag = NULL;
+			len = net_buf_findcrlf(rx_buf, &frag, &offset);
+			if (!frag) {
+				break;
+			}
+
+			/* look for matching data handlers */
+			i = -1;
+			for (i = 0; i < ARRAY_SIZE(handlers); i++) {
+				if (match(rx_buf, handlers[i].cmd) >= 0) {
+
+					/* found a matching handler */
+					LOG_INF("MATCH %s (len:%u)\n",
+						handlers[i].cmd, len);
+
+					/* skip cmd_len */
+					if (handlers[i].skip) {
+						rx_buf = net_buf_skip(rx_buf,
+						handlers[i].cmd_len);
+					}
+
+					/* locate next cr/lf */
+					frag = NULL;
+					len = net_buf_findcrlf(rx_buf,
+							       &frag, &offset);
+
+					/* call handler */
+					if (handlers[i].func) {
+						handlers[i].func(&rx_buf, len);
+					}
+
+					frag = NULL;
+					/* make sure buf still has data */
+					if (!rx_buf) {
+						break;
+					}
+
+					/* locate next cr/lf */
+					len = net_buf_findcrlf(rx_buf,
+							       &frag, &offset);
+					if (!frag) {
+						break;
+					}
+
+					break;
+				}
+			}
+
+			if (frag && rx_buf) {
+				/* clear out processed line (buffers) */
+				while (frag && rx_buf != frag) {
+					rx_buf = net_buf_frag_del(NULL, rx_buf);
+				}
+
+				net_buf_pull(rx_buf, offset);
+			}
+		}
+
+		/* give up time if we have a solid stream of data */
+		k_yield();
+	}
+}
+
+#if defined(CONFIG_WIFI_ESP8266_HAS_ENABLE_PIN)
+static void esp8266_gpio_reset(void)
+{
+	struct device *gpio_dev =
+		device_get_binding(CONFIG_WIFI_ESP8266_GPIO_DEVICE);
+
+	if (!gpio_dev) {
+		LOG_ERR("gpio device is not found: %s",
+			CONFIG_WIFI_ESP8266_GPIO_DEVICE);
+	}
+
+	gpio_pin_configure(gpio_dev, CONFIG_WIFI_ESP8266_GPIO_ENABLE_PIN,
+			   GPIO_DIR_OUT);
+
+	/* disable device until we want to configure it*/
+	gpio_pin_write(drv_data->gpio_dev, CONFIG_WIFI_ESP8266_GPIO_ENABLE_PIN,
+		       0);
+
+	/* enable device and check for ready */
+	k_sleep(100);
+	gpio_pin_write(drv_data->gpio_dev, CONFIG_WIFI_ESP8266_GPIO_ENABLE_PIN,
+		       1);
+}
+#endif
+
+static void esp8266_reset_work(void)
+{
+	int ret = -1;
+	int retry_count = 3;
+
+#if defined(CONFIG_WIFI_ESP8266_HAS_ENABLE_PIN)
+	esp8266_gpio_reset();
+#else
+	/* send AT+RST command */
+	while (retry_count-- && ret < 0) {
+		k_sleep(K_MSEC(100));
+		ret = send_at_cmd(NULL, "AT+RST", MDM_CMD_TIMEOUT);
+		if (ret < 0 && ret != -ETIMEDOUT) {
+			break;
+		}
+	}
+
+	if (ret < 0) {
+		LOG_ERR("cannot send reset %d\n", retry_count);
+		return;
```

Comment:
Not sure that all these **returns** will give you the outcome you are looking for.

You maybe need to consider a "goto error;" type idea, could have a reset label too.

Not great to use goto but difficult when using a-sync serial data.